[WIP] Add _sample_python methods

ritesh99rakesh opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Refer discussion https://github.com/sympy/sympy/issues/16942
Refer PR (closed) https://github.com/sympy/sympy/pull/17033

#### Brief description of what is fixed or changed
Add `_sample_python` methods for distributions

#### Other comments
TODO:
- [x] Add `_sample_python` methods for `crv_types`
- [x] Add `_sample_python` methods for `drv_types` (#17257 )
- [x] Add `_sample_python` methods for `frv_types` (#17445)

Ping @Upabjojr @sidhantnagpal @czgdp1807 

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

ritesh99rakesh committed:
Added _sample_python for NormalDistribution

ritesh99rakesh commented:
@Upabjojr Please review this PR.
I have added tests for `_sample_python` methods for `NormalDistribution`. If you want implementation in different form please let me know. By the time, I am adding `_sample_python` function for some other distributions in `crv_types.py`

czgdp1807 commented:
Run tests locally via,
```python
>>> import sympy
>>> sympy.test('stats/')
```

ritesh99rakesh committed:
Added _sample_python for GammaDistribution

czgdp1807 commented on sympy/stats/crv_types.py:
```diff
@@ -1379,8 +1379,24 @@ def pdf(self, x):
+    def _sample_scipy(self, scipy, size):
+        k, theta = float(self.k), float(self.theta)
+        from scipy.stats import gamma
+        return gamma.rvs(a=k, loc=0, scale=theta, size=size)
```
Is it possible that `theta` and `size` to be symbolic?

Upabjojr replied:
Well, if you are going to sample, I would say you need all variables to be numeric.

Upabjojr replied:
Do you see a way to numerically sample distributions with symbolic parameters? I think it could be done in certain cases, but I would recommend discussing this on gitter first.

ritesh99rakesh replied:
Yes, I agree with @Upabjojr that the variable should be numeric, but for some distributions like `NormalDistribution`, `GammaDistribution` etc we have used `return random.<distr>()` which allows for symbolic parameters but only gives one sample per call.
I have kept the function for cases where the sampling library is not mentioned and such function exist. Please tell if you want it removed.

czgdp1807 replied:
So, there are chances that parameter can be symbolic. I think we can add a check which can raise an informative message before passing them to `gamma.rvs`. Though we expect from the user to take care of that but may be if this feature is used in a large code base then may be some conditions may allow symbolic parameters to get passed to `gamma.rvs`. 
PS - This is just a suggestion. Infact, `gamma.rvs` will raise error if `theta` or `size` are symbolic, however that will not be very much informative.
You can do something like,
```python
if theta.has(Symbol) or size.has(Symbol):
    raise ValueError("Sampling is currently not possible for symbolic parameters.")
```

ritesh99rakesh replied:
@czgdp1807 this is the work of the dispatcher which will look for symbolic parameters. If the parameters are symbolic or no sample library is mentioned then `_sample_random` will be called (`_sample_random` will return an array if `random` could sample the distribution, otherwise will throw an error).
Please let me know what you think of this approach.

czgdp1807 replied:
Oh I see, no issues then. In fact, your approach will be better in avoiding the repetition of the checking code. From my side, this conversation can be resolved.

ritesh99rakesh commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -1205,3 +1205,35 @@ def test_conditional_eq():
     assert P(E > 1, Eq(E, 2)) == 1
     assert P(E < 1, Eq(E, 2)) == 0
+
+def test_temporary_sample_python():
```
I thought of the following:
> Please note that I have kept the tests as temporary since these functions should not be directly called. I will write a finalized test function after the dispatcher is ready.

Upabjojr replied:
The users shall not call these functions, but the tests can call them.

ritesh99rakesh commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -1031,7 +1031,7 @@ def test_density_unevaluated():
     assert isinstance(density(X+Y, evaluate=False)(z), Integral)
 
-
+@XFAIL # sample function now returns array
```
I also don't want to break the API but one change this PR introduces is that now every `_sample_python` function will return an array. Which will make `assert isinstance(nd.sample(), float) or nd.sample().is_Number` statement `False`.
I think that we should be returning an array since most of the time, we require samples > 1.

Upabjojr replied:
We are not going to change the current API. One possible solution is to pass the shape of the array to generate, so that `()` corresponds to scalars. `shape=()` could be made the default value.

ritesh99rakesh commented on sympy/stats/crv_types.py:
```diff
@@ -1379,8 +1379,24 @@ def pdf(self, x):
         k, theta = self.k, self.theta
         return x**(k - 1) * exp(-x/theta) / (gamma(k)*theta**k)
 
-    def sample(self):
```
Yes. This is a mistake and I will correct it.

Upabjojr replied:
No, it needs to be done now, otherwise the tests will not work anymore.

ritesh99rakesh committed:
Added _sample_python for ExponentialDistribution

ritesh99rakesh committed:
Corrected minor error and test

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -1379,8 +1395,23 @@ def pdf(self, x):
 
-    def sample(self):
-        return random.gammavariate(self.k, self.theta)
+    def _sample_random(self, size):
```
I would call this `_sample_python`, to make it clear that it uses the Python `random` standard library.

ritesh99rakesh committed:
Added for other distributions

ritesh99rakesh committed:
Added for some more distributions, corrected breaking cases

ritesh99rakesh commented:
@Upabjojr @sidhantnagpal I have updated the PR with above suggestions and also added `sample` methods for many `crv_types`.
Few questions:
1. Since some functions have sampling function in scipy but not in numpy and pymc3, I have added `_sample_scipy` for those. Should I write separate test cases for those?
2. Since we are keeping numeric sampling, for now, I have removed `_sample_python` function. What are your suggestions for this, should I remove or keep it?
3. Please comment on the current design of the code. If this is not good then I can change it. I will wait for your input before proceeding.

Upabjojr commented on sympy/stats/crv.py:
```diff
@@ -171,6 +172,8 @@ def check(*args):
 
     def sample(self):
         """ A random realization from the distribution """
+        scipy = import_module('scipy')
```
no imports here

Upabjojr commented on sympy/stats/crv.py:
```diff
@@ -171,6 +172,8 @@ def check(*args):
     def sample(self):
         """ A random realization from the distribution """
+        scipy = import_module('scipy')
+        return self._sample_scipy(scipy, size=())
```
call `_sample_python`

Upabjojr commented on sympy/stats/crv.py:
```diff
@@ -171,6 +172,8 @@ def check(*args):
         """ A random realization from the distribution """
+        scipy = import_module('scipy')
+        return self._sample_scipy(scipy, size=())
         icdf = self._inverse_cdf_expression()
```
this should go into `_sample_python`

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -172,6 +172,11 @@ def set(self):
     def pdf(self, x):
         return 1/(pi*sqrt((x - self.a)*(self.b - x)))
 
+    def _sample_scipy(self, scipy, size):
```
no `scipy` parameter

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -339,8 +344,20 @@ def pdf(self, x):
 
-    def sample(self):
-        return random.betavariate(self.alpha, self.beta)
+    def _sample_numpy(self, numpy, size):
```
no `numpy` parameter

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -339,8 +344,20 @@ def pdf(self, x):
         return x**(alpha - 1) * (1 - x)**(beta - 1) / beta_fn(alpha, beta)
 
-    def sample(self):
-        return random.betavariate(self.alpha, self.beta)
```
where did this code go?

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -339,8 +344,20 @@ def pdf(self, x):
+        from scipy.stats import beta
+        return beta.rvs(a=a, b=b, size=size)
+
+    def _sample_pymc3(self, pymc3, size):
```
no `pymc3` parameter

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -655,17 +688,20 @@ def check(k):
+
     def _characteristic_function(self, t):
         k = self.k
-
```
keep this newline

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -1033,8 +1084,20 @@ def check(rate):
         return self.rate * exp(-self.rate*x)
 
-    def sample(self):
-        return random.expovariate(self.rate)
```
where did this one go?

Upabjojr commented on sympy/stats/tests/test_rv.py:
```diff
@@ -117,7 +117,7 @@ def test_H():
     assert comp(H(G).evalf().round(2), 1.39)
 
-
+@XFAIL # sample function now returns array (currently for crv)
```
remove this line immediately, I don't want to see any `@XFAIL` added to the tests. If you need to add `@XFAIL`, it means your PR will cause damage to SymPy.

ritesh99rakesh commented:
@Upabjojr I have following queries so that I could work in the right direction:
1. Are we allowing for symbolic parameter now? If yes, then I will undo the delete of `random.<distr>` sample functions, otherwise they allow for only a few distributions and return one sample per call.
2. You mentioned `no numpy parameter` here (https://github.com/sympy/sympy/pull/17057#discussion_r296923670), but then how are we bringing we importing the `numpy` library.
3. Could you please tell the exact function of `_sample_python` and what will be its role in sampling.
4. At present this is what I understand the flow would be:
    - User calls the function `sample` with a library as the parameter.
    - `sample` function of the superclass gets executed which also decides which also imports the library before calling the necessary `sample` function (eg `_sample_numpy`, `_sample_scipy` etc) passing the import to it.
    - The corresponding `sample` function will return the sample.
If there is any change or addition required, please let me know.

Upabjojr commented:
> 1. Are we allowing for symbolic parameter now? If yes, then I will undo the delete of `random.<distr>` sample functions, otherwise they allow for only a few distributions and return one sample per call.

Let's put it into this perspective: why do you want to remove working code? I will never merge a PR removing working code from SymPy.

> 2. You mentioned `no numpy parameter` here ([#17057 (comment)](https://github.com/sympy/sympy/pull/17057#discussion_r296923670)), but then how are we bringing we importing the `numpy` library.

So you need to use `numpy` in a function. In order to do this, you import `numpy` in another function and then you pass it as a parameter to the function where it's needed. Why not import `numpy` in the function where it's needed?

> 3. Could you please tell the exact function of `_sample_python` and what will be its role in sampling.

`_sample_python` is the new name for `sample`. The function `sample` will be overwritten.

> `sample` function of the superclass gets executed which also decides ~which also imports the library before~ calling the necessary `sample` function (eg `_sample_numpy`, `_sample_scipy` etc) ~passing the import to it.~

Module are not passed as parameters to methods/functions.

Upabjojr commented:
Now that the workload is clear, can you please update this PR?

ritesh99rakesh committed:
Corrected crv samples

ritesh99rakesh commented:
@sidhantnagpal @Upabjojr I have updated the PR. Please review

ritesh99rakesh commented:
> > 1. You mentioned `no numpy parameter` here ([#17057 (comment)](https://github.com/sympy/sympy/pull/17057#discussion_r296923670)), but then how are we bringing we importing the `numpy` library.
> 
> So you need to use `numpy` in a function. In order to do this, you import `numpy` in another function and then you pass it as a parameter to the function where it's needed. Why not import `numpy` in the function where it's needed?
> 
I have a suggestion: For `crv_types` we are calling `_sample_*` function from `crv.py` under class `SingleContinuousDistribution`. If we import module there and then pass it as a parameter, we can avoid code duplication for again importing module for each `crv_type`. What do you think?

czgdp1807 commented on sympy/stats/crv_types.py:
```diff
@@ -341,8 +350,37 @@ def pdf(self, x):
-    def sample(self):
-        return random.betavariate(self.alpha, self.beta)
+    def _sample_python(self, size):
+        if size == ():
```
`if not size`? sounds more intuitive to me.

ritesh99rakesh replied:
Yes, thanks!

Upabjojr replied:
`if not size` is able to capture cases such as `size=[]`.

Also consider `if len(size) == 0:`, despite being longer and less pythonic, it has the advantage of being more readable (readers understand immediately that it is a tuple or array).

czgdp1807 commented:
Can something like the following be done?
Add some helper methods and use it in various classes. In other words, this can act as the lowest level of API,
_sample_helper_scipy, _sample_helper_numpy, . . . < _sample_scipy, _sample_pymc3,... < sample.
**Example**
```python
def _sample_helper_scipy(self, size, distribution, *args):
        args = map(float, *args)
        scipy = import_module(module)
        if scipy:
            dist = getattr(__import__('scipy.stats', globals(), locals(), [distribution], 0), distribution)
            return dist.rvs(*args) # pass size in args directly.
        else:
            raise NotImplementedError('Sampling the Arcsin Distribution requires Scipy.')
```
Probably this way you can avoid repition and reduce the size of the code. Rest depends on the mentors, if they want to allow this kind of thing. Thanks.

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -2590,14 +2809,32 @@ def pdf(self, x):
-    def sample(self):
+    def _sample_numpy(self, size):
+        xm, alpha = float(self.xm), float(self.alpha)
+        numpy = import_module('numpy')
```
what about importing `numpy` globally?

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -2590,14 +2809,32 @@ def pdf(self, x):
+        if numpy:
+            return (numpy.random.pareto(a=alpha, size=1000) + 1) * xm
+        else:
+            raise NotImplementedError('Sampling the Pareto Distribution requires Numpy.')
```
I would remove this `if numpy:` statement... just `return ...` without any exception raising.

Upabjojr commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -1240,3 +1240,36 @@ def test_conditional_eq():
+    for X in distribs:
+        numpy = import_module('numpy')
+        if not numpy:
+            skip('Numpy not installed. Abort tests for _sample_numpy.')
```
maybe you should only skip the `numpy` tests? Same for the other modules.

Upabjojr commented on sympy/stats/crv.py:
```diff
@@ -169,8 +170,9 @@ def __new__(cls, *args):
-    def sample(self):
+    def sample(self, size=()):
         """ A random realization from the distribution """
+        return self._sample_scipy(size)
```
This is supposed to be `_sample_python` for now.

Upabjojr commented on sympy/stats/crv.py:
```diff
@@ -169,8 +170,9 @@ def __new__(cls, *args):
+    def sample(self, size=()):
         """ A random realization from the distribution """
+        return self._sample_scipy(size)
         icdf = self._inverse_cdf_expression()
```
code after return is dead. Remove it.

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -341,8 +350,37 @@ def pdf(self, x):
+    def _sample_python(self, size):
+        if size == ():
+            return random.betavariate(self.alpha, self.beta)
+        return [random.betavariate(self.alpha, self.beta) for i in range(size)]
```
what if `size` is multidimensional? For example, you want to generate a matrix?

ritesh99rakesh replied:
multidimensional `size` is handled by all `exporting` libraries, for `_sample_python` I have to write code.

Upabjojr replied:
Well... we have created `ArrayComprehension` at the beginning of this summer. It's meant to handle these kinds of cases.

ritesh99rakesh replied:
Right! Thanks!!

Upabjojr replied:
`ArrayComprehension` with `lambda`? `ArrayComprehension(lambda: random.betavariate(1, 1), ...)` ?

Upabjojr commented:
I would recomment to make imports at the module level, i.e. `numpy = import_module("numpy")` in the main scope.

`_sample_numpy(...)` would just call `numpy. <...>`. We don't need to handle the exception, in the future `.sample( )` will properly wrap the call.

ritesh99rakesh committed:
Remove unnecessary code

ritesh99rakesh committed:
Sample ContinuousHandmadeDistribution

Upabjojr commented on sympy/stats/crv.py:
```diff
@@ -323,6 +324,22 @@ class ContinuousDistributionHandmade(SingleContinuousDistribution):
+            def _pdf(self, z):
+                return handmade_pdf(z)
+
+        scipy_rv = scipy_pdf(a=self.set._inf, b=self.set._sup, name='scipy_pdf')
```
nice idea +1

Upabjojr commented on sympy/stats/crv.py:
```diff
@@ -323,6 +324,22 @@ class ContinuousDistributionHandmade(SingleContinuousDistribution):
+
+    def sample(self, size=()):
+        icdf = self._inverse_cdf_expression()
+        return icdf(random.uniform(0, 1))
```
This should be `_sample_python`. `sample` should simply link to `_sample_python`.

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -176,6 +180,15 @@ def set(self):
+        a = float(self.a)
+        scipy = import_module('scipy')
+        if scipy:
+            from scipy.stats import arcsine
```
`scipy` is already imported. Have you tried `arcsine = scipy.stats.arcsine` ?

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -176,6 +180,15 @@ def set(self):
+            from scipy.stats import arcsine
+            return arcsine.rvs(loc=a, size=size)
+        else:
+            raise NotImplementedError('Sampling the Arcsin Distribution requires Scipy.')
```
No need for this.

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -176,6 +180,15 @@ def set(self):
+    def _sample_scipy(self, size):
+        a = float(self.a)
+        scipy = import_module('scipy')
+        if scipy:
```
No need for this. It makes the code too long. The check about scipy's existence will be put somewhere else in the future. Just assume `scipy` is imported.

ritesh99rakesh committed:
Minor change

ritesh99rakesh commented:
@Upabjojr I have updated the PR accordingly. I have also stated my issue with using `ArrayComprehension`. Please review.

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -176,6 +180,11 @@ def set(self):
+    def _sample_scipy(self, size):
+        a = float(self.a)
+        from scipy.stats import arcsine
+        return arcsine.rvs(loc=a, size=size)
```
What about `scipy.stats.arcsine` ? You don't need to import twice.

ritesh99rakesh commented:
@Upabjojr the following does not work:
```python3
>>> ArrayComprehension((lambda: random.betavariate(1, 1))(), (i, 1, 4)).doit()
[0.262993991747324, 0.262993991747324, 0.262993991747324, 0.262993991747324]
```
Reason:
```python3
>>> ArrayComprehension((lambda: random.betavariate(1, 1))(), (i, 1, 4)).function
0.1234487433090109
>>> ArrayComprehension((lambda i: random.betavariate(1, 1))(i), (i, 1, 4)).function
0.5671487433490109
```
What do you suggest should I do?

Upabjojr commented:
> What do you suggest should I do?

@kangzhiq hi, there is a problem with `ArrayComprehension`. I believe it does not support function calling. Can you have a look at it?

kangzhiq commented:
@Upabjojr @ritesh99rakesh 
Please feel free to comment on the new PR about ArrayComprehension: #17408

ritesh99rakesh committed:
Complete all continuous distribution sample methods

Smit-create commented:
I would like to continue this PR once #18289 gets merged

czgdp1807 commented:
@Smit-create The said PR has been merged. You can continue this now.

Smit-create commented:
Yes I would start working on this and submit a PR within few days. Thanks

czgdp1807 commented:
Closing in favour of https://github.com/sympy/sympy/pull/18754
